File: Examples/test6.py
from dataclasses import dataclass

type MyList[A] = MyNil[A] | MyCons[A]


@dataclass
class MyNil[A]:
    pass


@dataclass
class MyCons[A]:
    hd: A
    tl: MyList[A]


type Colour = Red | Black | White


@dataclass
class Red:
    pass


@dataclass
class Black:
    pass


@dataclass
class White:
    pass


type Tree[A, C] = Leaf[A, C] | LNode[A, C]


@dataclass
class Leaf[A, C]:
    value: A


@dataclass
class LNode[A, C]:
    value: C
    left: Tree[A, C]
    right: Tree[A, C]


def toInt(c: Colour) -> int:
    match c:
        case Red():
            return 0
        case Black():
            return 1
        case White():
            return 2


def length[A](l: MyList[A]) -> int:
    match l:
        case MyNil():
            return 0
        case MyCons(hd=_, tl=t):
            return 1 + length(t)


def max(x: int, y: int) -> int:
    if x > y:
        return x
    else:
        return y


def size[A, B](t: Tree[A, B]) -> int:
    match t:
        case Leaf(value=_):
            return 1
        case LNode(value=_, left=l, right=r):
            return max(size(l), size(r))


c = Red()
l: Tree[int, str] = Leaf(value=3)
t = LNode(value="abc", left=Leaf(value=7), right=Leaf(value=size(l)))

File: Examples/TS/test6.ts
type MyList<A> = MyNil<A> | MyCons<A>;
interface MyNil<A> {
}
interface MyCons<A> {
    hd: A;
    tl: MyList<A>;
}
type Colour = Red | Black | White;
interface Red{
}
interface Black{
}
interface White{
}
type Tree<A, C> = Leaf<A, C> | LNode<A, C>;
interface Leaf<A, C> {
    value: A;
}
interface LNode<A, C> {
    value: C;
    left: Tree<A, C>;
    right: Tree<A, C>;
}
function toInt(c: Colour): number {
    if (c.kind === "Red") {
        const  = c.value;
        
        return 0;
    } else if (c.kind === "Black") {
        const  = c.value;
        
        return 1;
    } else if (c.kind === "White") {
        const  = c.value;
        
        return 2;
    }
}
function length<A>(l: MyList<A>): number {
    if (l.kind === "MyNil") {
        const  = l.value;
        
        return 0;
    } else if (l.kind === "MyCons") {
        const hd: _, tl: t = l.value;
        
        return (1 + length(t));
    }
}
function max(x: number, y: number): number {
    if (x > y) {
        return x;
    } else {
        return y;
    }
}
function size<A, B>(t: Tree<A, B>): number {
    if (t.kind === "Leaf") {
        const value: _ = t.value;
        
        return 1;
    } else if (t.kind === "LNode") {
        const value: _, left: l, right: r = t.value;
        
        return max(size(l), size(r));
    }
}
const c = { kind: "Red", value: {  } }
const l: Tree<number, string> = { kind: "Leaf", value: { value: 3 } }
const t = { kind: "LNode", value: { value: "abc", left: { kind: "Leaf", value: { value: 7 } }, right: { kind: "Leaf", value: { value: size(l) } } } }
